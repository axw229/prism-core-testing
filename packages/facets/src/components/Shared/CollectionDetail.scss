.collection-detail {
  &__wrapper {
    display: flex;
    background: var(getThemeColor('primaryBg'));
    height: 100%;
    flex-direction: column;
    @include respond-above(sm) {
      flex-direction: row;
    }
  }
  &__info {
    display: block;
    float: left;
    height: 100%;
    width: 100%;
    text-align: center;
    @include respond-above(sm) {
      width: 34%;
    }
  }
  &__cover {
    width: 100%;
    &--hidden-mobile {
      @include respond-below(sm) {
        display: none;
      }
    }
  }
  &__description {
    padding: 1em 2em;
    text-align: left;
    font-family: var(getVarName('typography.bodyFontFamily'));
    white-space: pre-wrap;
    &--hidden-mobile {
      @include respond-below(sm) {
        display: none;
      }
    }
  }
  &__border { margin: 1rem 0 3rem }
  &__button {
    color: var(getThemeColor('buttonColor'));
    background-color: var(getThemeColor('buttonBgColor'));
    border: 1px solid var(getThemeColor('buttonBorder'));
    text-transform: var(getVarName('typography.buttonTextTransform'));
    font-weight: var(getVarName('typography.buttonFontWeight'));
    padding: 10px 25px;
    font-size: 0.9em;
    transition: all .2s;
    position: relative;
    text-decoration: none;
    @include respond-below(sm) {
      margin-bottom: 20px;
    }

    &:hover {
      border-color: var(getThemeColor('buttonHoverBorder'));
      background-color: var(getThemeColor('buttonHoverBgColor'));
      color: var(getThemeColor('buttonHoverColor'));
    }

    &__loader {
      transition: all .2s;
      position: absolute;
      font-size: 2em;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      justify-content: center;
      background: var(getThemeColor('buttonBgColor.trans'));
    }
  }
  &__learn-more {
    display: inline-block;
    text-align: center;
    text-transform: uppercase;
    width: 100%;
    background: 0 0;
    color: var(getThemeColor('secondary.darker'));
    fill: #000;
    padding: .75em 1.75em;
    font-size: .8em;
    &:hover {
      color: var(getThemeColor('secondary'));
      border-color: var(getThemeColor('secondary'));
      text-decoration: none;
      cursor: pointer;
    }
    @include respond-above(sm) {
      width: auto;
      border: 1px solid var(getThemeColor('light.lightest'));
      background: var(getThemeColor('white'));
    }
  }
  &__color-list {
    float: right;
    width: 100%;
    height: 400px;
    position: relative;
    overflow: hidden;

    @include respond-above(sm) {
      width: 66%;
      height: 535px;
    }
  }
  &__color-list-vertical { height: inherit }
  &__trigger {
    border: 1px solid var(getThemeColor('light.lightest'));
    display: inline-block;
    text-align: center;
    text-transform: uppercase;
    width: auto;
    color: var(getThemeColor('black'));
    fill: var(getThemeColor('black'));
    height: auto;
    padding: .3rem .5rem;
    transition: all .3s ease;
    background: var(getThemeColor('white.trans'));
    cursor: pointer;
    position: absolute;
    z-index: 1;
    left: calc(50% - 42px/2 - 16px/2);
    &:hover {
      color: var(getThemeColor('secondary'));
      border-color: var(getThemeColor('secondary'));
    }
  }
  &__previous-trigger {
    top: 0;
  }
  &__next-trigger {
    bottom: 0;
  }

  &--house-shaped {
    height: 108px !important;
    margin-top: 48px !important;
    padding: 0 !important;
    box-shadow: none;
    outline: none;
    transition: none;

    &::before {
      content: '';
      display: block;
      position: absolute;
      top: -47%;
      width: 100%;
      height: 47.5%;
      background-color: inherit;
      clip-path: polygon(50% 0, 0 100%, 100% 100%);
      cursor: default;
    }
  }
}
