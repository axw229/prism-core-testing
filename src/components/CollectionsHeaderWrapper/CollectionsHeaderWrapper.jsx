// @flow
import React from 'react'
import './CollectionsHeaderWrapper.scss'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

type Props = {
}

type State = {
  isShowBack: boolean,
  header: string
}

const baseClass = 'collections-header'

const CollectionsHeaderWrapper = (WrappedComponent: any) => {
  class CollectionsHeader extends React.Component<Props, State> {
    constructor (props) {
      super(props)
      this.state = {
        isShowBack: false,
        header: ''
      }
      this.showBack = this.showBack.bind(this)
      this.backHandler = this.backHandler.bind(this)
      this.setHeader = this.setHeader.bind(this)
    }

    showBack () {
      this.setState({
        isShowBack: true
      })
    }

    backHandler () {
      this.setState({
        isShowBack: false
      })
    }

    setHeader (header) {
      this.setState({
        header
      })
    }

    render () {
      const { isShowBack, header } = this.state

      return (
        <div className={`${baseClass}__wrapper`}>
          <div className={`${baseClass}__header`}>
            <div className={`${baseClass}__heading`}>{header}</div>
            {(isShowBack) ? <button className={`${baseClass}__button ${baseClass}__button--left`} onClick={this.backHandler}>
              <div><FontAwesomeIcon className={``} icon={['fa', 'angle-left']} />&nbsp;<span className={`${baseClass}__button-left-text`}>BACK</span></div>
            </button> : ''}
            <button className={`${baseClass}__button ${baseClass}__button--right`}>
              <div className={`${baseClass}__close`}><span>CLOSE</span>&nbsp;<FontAwesomeIcon className={``} icon={['fa', 'chevron-up']} /></div>
              <div className={`${baseClass}__cancel`}><FontAwesomeIcon className={``} icon={['fa', 'times']} /></div>
            </button>
          </div>
          <div className={`${baseClass}__content`}>
            <WrappedComponent {...this.props} showBack={this.showBack} isShowBack={isShowBack} setHeader={this.setHeader} />
          </div>
        </div>
      )
    }
  }

  return CollectionsHeader
}

export default CollectionsHeaderWrapper
