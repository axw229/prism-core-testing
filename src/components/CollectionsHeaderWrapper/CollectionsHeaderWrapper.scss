.collections-header {
  &__wrapper {
    font-size: 1em;
    font-weight: 400;
    color: #333;
    line-height: 1.2em;
    height: 100%;
    user-select: none;
    background: #f0f0f0;
    width: 100%;
    max-width: 100%;
  }
  &__content {
    display: block;
  }
  &__header {
    display: block;
    position: relative;
  }
  &__heading {
    font-size: 1.125em;
    line-height: 1.350em;
    background: #fff;
    display: block;
    overflow: hidden;
    padding: 3% 16% 1.25em;
    position: relative;
    text-overflow: ellipsis;
    white-space: nowrap;
    z-index: 2;
    text-align: center;
    @include respond-above(sm) {
      background: #f0f0f0;
    }
  }
  &__button {
    background: none;
    border: none;
    display: inline-block;
    text-align: center;
    text-transform: uppercase;
    width: auto;
    color: #000;
    fill: #000;
    padding: 0;
    top: .875rem;
    position: absolute;
    z-index: 3;
    &:hover {
      border-color: #2cabe2;
      cursor: pointer;
      color: #2cabe2;
    }
    &:focus {
      outline-style: solid;
      outline-color: #2cabe2;
      outline-width: 2px;
      z-index: 100;
    }
    &--left {
      left: .875rem;
    }
    &--right {
      right: .875rem;
    }
    @include respond-above(sm) {
      background: #fff;
      border: 1px solid #f2f2f2;
      padding: .75em 1.75em;
    }
  }
  &__button-left-text {
    display: none;
    text-transform: uppercase;
    @include respond-above(sm) {
      display: inline-block;
    }
  }
  &__close {
    display: none;
    @include respond-above(sm) {
      display: block;
    }
  }
  &__cancel {
    display: block;
    @include respond-above(sm) {
      display: none;
    }
  }
}