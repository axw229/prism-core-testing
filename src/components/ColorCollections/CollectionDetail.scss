.collection-detail {
  &__wrapper {
    display: flex;
    background: #f0f0f0;
    height: 100%;
    flex-direction: column;
    @include respond-above(sm) {
      flex-direction: row;
    }
  }
  &__info {
    display: block;
    float: left;
    height: 100%;
    width: 100%;
    text-align: center;
    @include respond-above(sm) {
      width: 34%;
    }
  }
  &__cover {
    display: none;
    width: 100%;
    @include respond-above(sm) {
      display: block;
    }
  }
  &__description {
    padding: 1em 2em;
    text-align: left;
    display: none;
    @include respond-above(sm) {
      display: block;
    }
  }
  &__learn-more {
    display: inline-block;
    text-align: center;
    text-transform: uppercase;
    width: 100%;
    background: 0 0;
    color: #0069af;
    fill: #000;
    padding: .75em 1.75em;
    font-size: .8em;
    &:hover {
      color: #2cabe2;
      border-color: #2cabe2;
      text-decoration: none;
      cursor: pointer;
    }
    @include respond-above(sm) {
      width: auto;
      border: 1px solid #f2f2f2;
      background: #fff;
    }
  }
  &__color-list {
    float: right;
    width: 100%;
    position: relative;
    overflow: scroll;
    @include respond-above(sm) {
      width: 66%;
    }
  }
  &__color-list-vertical {
    height: 58vw;
    max-height: 500px;
    white-space: wrap;
  }
  &__trigger {
    border: 1px solid #f2f2f2;
    display: inline-block;
    text-align: center;
    text-transform: uppercase;
    width: auto;
    color: #000;
    fill: #000;
    height: auto;
    padding: 1em;
    transition: all .3s ease;
    background: rgba(255,255,255,.85);
    cursor: pointer;
    position: absolute;
    z-index: 1;
    left: calc(50% - 42px/2 - 16px/2);
    &:hover {
      color: #2cabe2;
      border-color: #2cabe2;
    }
  }
  &__previous-trigger {
    top: 0;
  }
  &__next-trigger {
    bottom: 0;
  }
}