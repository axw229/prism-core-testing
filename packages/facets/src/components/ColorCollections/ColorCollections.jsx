// @flow
import React, { useCallback, useContext, useEffect, useMemo, useState } from 'react'
import { useIntl } from 'react-intl'
import { shallowEqual, useDispatch, useSelector } from 'react-redux'
import * as GA from 'src/analytics/GoogleAnalytics'
import CardMenu from 'src/components/CardMenu/CardMenu'
import Carousel from 'src/components/Carousel/Carousel'
import ColorStripButton from 'src/components/ColorStripButton/ColorStripButton'
import CollectionDetail from 'src/components/Shared/CollectionDetail'
import ColorCollectionsTab from 'src/components/Shared/ColorCollectionsTab'
import { GA_TRACKER_NAME_BRAND } from 'src/constants/globals'
import ConfigurationContext, {
  type ConfigurationContextType
} from 'src/contexts/ConfigurationContext/ConfigurationContext'
import { loadCollectionSummaries } from 'src/store/actions/collectionSummaries'
import { loadColors } from 'src/store/actions/loadColors'
import './ColorCollections.scss'

export function ColorCollections() {
  const dispatch = useDispatch()
  const { formatMessage, locale } = useIntl()
  const { brandId, cvw = {} } = useContext<ConfigurationContextType>(ConfigurationContext)
  const { colorCollections = {} } = cvw
  const { pagerPosition = 'center', subtitle } = colorCollections

  const { summaries, categories } = useSelector((state) => state.collectionSummaries, shallowEqual)
  const colorMap = useSelector((state) => state.colors.items.colorMap, shallowEqual)
  useEffect(() => {
    loadCollectionSummaries(brandId, { language: locale })(dispatch)
  }, [])
  useEffect(() => {
    loadColors(brandId)(dispatch)
  }, [])

  const [tabId, setTabId] = useState()
  const [collectionData, setCollectionData] = useState([])

  useEffect(() => {
    if (typeof tabId !== 'undefined') {
      const category = categories.data.filter(({ id }) => `${id}` === `${tabId}`)[0]
      const collectionData =
        colorMap && category
          ? category.summaryIds.map((summaryId) => {
              const { colorIds, coverUrl, description, name, pdfUrl, thumbColorIds, thumbUrl } =
                summaries.data[summaries.idToIndexHash[summaryId]]
              return {
                collections: colorIds.map((id) => colorMap[id]),
                coverUrl,
                description,
                name,
                pdfUrl,
                thumbColors: thumbColorIds.map((id) => colorMap[id]),
                thumbUrl
              }
            })
          : []
      setCollectionData(collectionData)
    }
  }, [tabId, colorMap])

  useEffect(() => {
    if (categories.data.length) {
      setTabId(categories.data[0].id)
    }
  }, [categories])

  return (
    <CardMenu menuTitle={formatMessage({ id: 'COLOR_COLLECTIONS' })}>
      {(setCardShowing, setCardTitle) => (
        <>
          {subtitle && <div className='color-collections__subtitle'>{subtitle}</div>}
          <div className='color-collections__wrapper'>
            <ColorCollectionsTab collectionTabs={categories.data} showTab={setTabId} tabIdShow={tabId} />
            <div className='color-collections__collections-list' role='main'>
              <Carousel
                BaseComponent={ColorStripButtonWrapper}
                btnRefList={[]}
                data={collectionData}
                defaultItemsPerView={8}
                isInfinity={false}
                getSummaryData={(collectionSummaryData) => {
                  setCardShowing(<CollectionDetail collectionDetailData={collectionSummaryData} />)
                  setCardTitle(collectionSummaryData.name)
                  GA.event(
                    { category: 'Color Collections', action: 'Collection Click', label: collectionSummaryData.name },
                    GA_TRACKER_NAME_BRAND[brandId]
                  )
                }}
                key={tabId}
                pagerPosition={pagerPosition}
                showPageIndicators={collectionData.length > 8}
              />
            </div>
          </div>
        </>
      )}
    </CardMenu>
  )
}

const ColorStripButtonWrapper = (props: any) => {
  const { data, getSummaryData, itemNumber, btnRefList, onKeyDown } = props
  const clickHandler = useCallback(() => getSummaryData(data), [data])
  const colors = useMemo(() => data.thumbColors.slice(0, 5), [data.thumbColors])
  btnRefList[itemNumber] = React.useRef()
  let imgAltText = `${data.name}.`

  if (colors.length > 0) {
    imgAltText += ' Color group includes '
    colors.map((color, index) => {
      imgAltText += colors.length === index + 1 ? `${color.name}.` : `${color.name}, `
    })
  }

  return (
    <ColorStripButton
      bottomLabel={data.name}
      colors={colors}
      onClick={clickHandler}
      onKeyDown={onKeyDown}
      ref={btnRefList[itemNumber]}
    >
      <img className='collection__summary__top-section__image' alt={imgAltText} src={data.thumbUrl} />
    </ColorStripButton>
  )
}

export default ColorCollections
