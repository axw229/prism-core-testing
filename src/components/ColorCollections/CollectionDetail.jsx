// @flow
import React, { useRef } from 'react'
import { connect } from 'react-redux'
// $FlowIgnore -- no defs for react-virtualized
import { Grid, AutoSizer } from 'react-virtualized'
import ColorWallSwatch from '../Facets/ColorWall/ColorWallSwatch/ColorWallSwatch'
import ColorWallContext from '../Facets/ColorWall/ColorWallContext'
import { add } from '../../store/actions/live-palette'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import * as scroll from 'scroll'
import './CollectionDetail.scss'
import type { ColorCollectionDetail, Color } from '../../shared/types/Colors'

const GRID_AUTOSCROLL_SPEED: number = 300

type Props = {
  collectionDetailData: ColorCollectionDetail,
  addToLivePalette: Function
}

type Ref = RefObject

const cwProviderValues = {
  displayDetailsLink: false,
  displayInfoButton: true,
  displayAddButton: true
}

const baseClass = 'collection-detail'
const wrapper = `${baseClass}__wrapper`
const collectionInfo = `${baseClass}__info`
const collectionCover = `${baseClass}__cover`
const collectionDescription = `${baseClass}__description`
const collectionLearnMore = `${baseClass}__learn-more`
const collectionColorList = `${baseClass}__color-list`
const collectionColorListVertical = `${baseClass}__color-list-vertical`
const collectionVerticalControls = `${baseClass}__vertical-controls`
const verticalControlsTrigger = `${baseClass}__trigger`
const triggerPrevious = `${baseClass}__previous-trigger`
const triggerNext = `${baseClass}__next-trigger`

function cellRenderer ({
  columnIndex, // Horizontal (column) index of cell
  isScrolling, // The Grid is currently being scrolled
  isVisible, // This cell is visible within the grid (eg it is not an overscanned cell)
  key, // Unique key within array of cells
  parent, // Reference to the parent Grid (instance)
  rowIndex, // Vertical (row) index of cell
  style // Style object to be applied to cell (to position it)
}: Object) {
  const { props: { columnCount, addToLivePalette, colors } } = parent
  const index = columnIndex + (rowIndex * columnCount)

  if (colors && colors[index]) {
    const thisColor: Color = colors[index]

    return (
      <div key={key} style={style}>
        <ColorWallContext.Provider value={cwProviderValues}>
          <ColorWallSwatch
            key={thisColor.hex}
            showContents
            color={thisColor}
            onAdd={addToLivePalette}
          />
        </ColorWallContext.Provider>
      </div>
    )
  }

  return null
}

let _gridHeight = 0
let _cellSize = 0
function handleGridResize ({ width, height }) {
  _gridHeight = height
}

export const CollectionDetail = (props: Props) => {
  const resultSwatchSize = 175
  const { addToLivePalette, collectionDetailData } = props
  const _gridWrapperRef: ?Ref = useRef(null)

  return (
    <div className={`${wrapper}`}>
      <div className={`${collectionInfo}`}>
        <img className={`${collectionCover}`} alt='' src={`${collectionDetailData.img}`} />
        <div className={`${collectionDescription}`}>{collectionDetailData.name}</div>
        <a href='https://' className={`${collectionLearnMore}`}>LEARN MORE ABOUT THIS COLLECTION</a>
      </div>
      <div className={`${collectionColorList}`}>
        <div ref={_gridWrapperRef} className={`${collectionColorListVertical}`}>
          <AutoSizer onResize={handleGridResize}>
            {({ height, width }) => {
              const columnCount = Math.max(1, Math.round(width / resultSwatchSize))
              const rowCount = Math.ceil(collectionDetailData.collections.length / columnCount)
              const newSize = width / columnCount
              _cellSize = newSize
              return (
                <Grid
                  colors={collectionDetailData.collections}
                  cellRenderer={cellRenderer}
                  columnWidth={newSize}
                  columnCount={columnCount}
                  height={height}
                  rowHeight={newSize}
                  rowCount={rowCount}
                  width={width}
                  addToLivePalette={addToLivePalette}
                />
              )
            }}
          </AutoSizer>
        </div>
        <div className={`${collectionVerticalControls}`}>
          <div role='presentation' onClick={() => verticalScroll(_gridWrapperRef, 'top', _gridHeight, _cellSize)} className={`${verticalControlsTrigger} ${triggerPrevious}`}>
            <FontAwesomeIcon className={``} icon={['fa', 'angle-up']} />
          </div>
          <div role='presentation' onClick={() => verticalScroll(_gridWrapperRef, 'bottom', _gridHeight, _cellSize)} className={`${verticalControlsTrigger} ${triggerNext}`}>
            <FontAwesomeIcon className={``} icon={['fa', 'angle-down']} />
          </div>
        </div>
      </div>
    </div>
  )
}

function verticalScroll (_gridWrapperRef, direction, _gridHeight, _cellSize) {
  if (_gridWrapperRef && _gridWrapperRef.current) {
    const gridEl = _gridWrapperRef.current.querySelector('.ReactVirtualized__Grid')

    let scrollToY = 0
    if (gridEl && gridEl.scrollTop) {
      scrollToY = gridEl.scrollTop
    }

    if (direction === 'bottom') {
      scrollToY += _cellSize
    } else {
      scrollToY -= _cellSize
    }
    scroll.top(gridEl, scrollToY, { duration: GRID_AUTOSCROLL_SPEED })
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    addToLivePalette: (color) => {
      dispatch(add(color))
    }
  }
}

export {
  collectionCover,
  collectionDescription,
  triggerPrevious,
  triggerNext
}
export default connect(null, mapDispatchToProps)(CollectionDetail)
