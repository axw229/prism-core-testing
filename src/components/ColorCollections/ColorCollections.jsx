// @flow
import CollectionDetail from './CollectionDetail'
import CollectionsHeaderWrapper from '../CollectionsHeaderWrapper/CollectionsHeaderWrapper'
import ColorCollectionsTab from './ColorCollectionsTab'
import ExpertColorDetails from '../Carousel/ExpertColorDetails'
import React, { useState, useEffect } from 'react'
import WithConfigurationContext from '../../contexts/ConfigurationContext/WithConfigurationContext'

import { ColorListWithCarousel } from '../Carousel/Carousel'
import { connect } from 'react-redux'
import { expertColorsData } from '../Carousel/data'
import { injectIntl } from 'react-intl'
import { loadCollectionSummaries as loadCS } from '../../store/actions/collectionSummaries'
import { loadColors } from '../../store/actions/loadColors'

import './ColorCollections.scss'

type Props = {
  categories: Object,
  colorMap: Object,
  config: Object,
  intl: Object,
  isExpertColor: boolean,
  isShowBack: boolean,
  loadColors: Function,
  loadCS: Function,
  setHeader: Function,
  showBack: Function,

}

const baseClass = 'color-collections'
const wrapper = `${baseClass}__wrapper`
export const collectionsList = `${baseClass}__collections-list`

ColorCollections.collectionData = []

ColorCollections.getSummary = function getSummary (id, props) {
  const {
    name,
    thumbUrl: img,
    description,
    colorIds
    // ...rest // has a bunch of stuff

  } = props.summaries.data[props.summaries.idToIndexHash[id]]

  return {
    collections: colorIds
      .map(id => props
        .colorMap[id]
      ).filter(collection => collection),
    description,
    img,
    name
  }
}

ColorCollections.getSummariesForTab = function getSummariesForTab (
  tabId, props
) {
  const category = props
    .categories
    .data[props.categories.idToIndexHash[tabId]]

  if (category) {
    return category.summaryIds.reduce(
      (summaries, summaryId) => {
        summaries.push(ColorCollections.getSummary(summaryId, props))
        return summaries
      },
      []
    )
  }

  return []
}

ColorCollections.updateCollectionData = function updateCollectionData ({
  data = expertColorsData,
  props = {},
  tabId
}) {
  ColorCollections.collectionData = (props.isExpertColor)
    ? data
    : ColorCollections.getSummariesForTab(tabId, props)
}

export function ColorCollections (props: Props) {
  const [hasColors, setHasColors] = useState(!!Object.keys(props.colorMap).length)

  // TODO:noah.hall
  // wtf do we show while loading?
  // check with cody for loading screen
  const [csLoaded, setCsLoaded] = useState(false)
  useEffect(() => {
    if (!csLoaded) {
      setCsLoaded(true)
      props.loadCS()
    }
  })

  const [colorsRequested, setColorsRequested] = useState(false)
  // load colors if dont exist
  if (!colorsRequested && !hasColors) {
    props.loadColors(props.config.brandId, { language: props.intl.locale })
    setColorsRequested(true)
  }

  const { isShowBack, showBack, setHeader, isExpertColor } = props
  const [tabIdShow, showTab] = useState('')
  const [collectionDataDetails, updateCollectionDataDetails] = useState({})

  const showTabHandler = (tabId: string, isClickTab: boolean) => {
    showTab((prevTab) => {
      if (prevTab !== tabId) ColorCollections.updateCollectionData({ tabId, props })

      return tabId
    })
  }

  // initial render, selected tab unknown
  // set selected tab
  if (!tabIdShow && props.categories.data.length) {
    showTabHandler(props.categories.data[0].id)
  }
  // subsequent renders, tab known
  // once colors exist, update collection only once
  if (tabIdShow && !hasColors && Object.keys(props.colorMap).length) {
    setHasColors(true)
    ColorCollections.updateCollectionData({ tabId: tabIdShow, props })
  }

  const headerContent = (isExpertColor)
    ? 'Expert Color Picks'
    : 'Color Collections'

  useEffect(() => {
    if (!isShowBack) {
      setHeader(headerContent)
    }
  }, [isShowBack])

  if (isShowBack === true) {
    return (isExpertColor)
      ? <ExpertColorDetails expertColors={collectionDataDetails} />
      : <CollectionDetail collectionDetailData={collectionDataDetails} />
  }

  const onClickHandler = (collectionSummaryData: Object) => {
    showBack()
    if (!isExpertColor) {
      setHeader(collectionSummaryData.name)
    }
    updateCollectionDataDetails(collectionSummaryData)
  }

  return (
    <div className={`${wrapper}`}>
      {
        (!isExpertColor) &&
          <ColorCollectionsTab
            collectionTabs={props.categories.data}
            showTab={showTabHandler}
            tabIdShow={tabIdShow}
          />
      }
      {
        <div className={`${collectionsList}`}>
          <ColorListWithCarousel
            defaultItemsPerView={8}
            isInfinity={false}
            key={tabIdShow}
            data={ColorCollections.collectionData}
            getSummaryData={onClickHandler}
            isExpertColor={isExpertColor}
          />
        </div>
      }
    </div>
  )
}

const mapStateToProps = (state) => {
  const {
    collectionSummaries: {
      categories,
      summaries
    },
    colors: {
      items: {
        colorMap = {}
      }
    }
  } = state

  return {
    categories,
    summaries,
    colorMap
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    loadCS () { dispatch(loadCS()) },
    loadColors (brandId, opts) { dispatch(loadColors(brandId, opts)) }
  }
}
export default CollectionsHeaderWrapper(connect(
  mapStateToProps,
  mapDispatchToProps
)(injectIntl(WithConfigurationContext(ColorCollections))))
