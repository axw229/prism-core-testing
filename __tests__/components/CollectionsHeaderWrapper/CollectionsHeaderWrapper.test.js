/* eslint-env jest */
import React from 'react'
import { shallow, mount } from 'enzyme'
import CollectionsHeaderWrapper, { buttonLeft, heading, buttonRight } from 'src/components/CollectionsHeaderWrapper/CollectionsHeaderWrapper'
import CollectionDetail from 'src/components/Shared/CollectionDetail'
import { ColorCollections } from 'src/components/ColorCollections/ColorCollections'
import { Link } from 'react-router-dom'

let defaultProps = {
}

const getCollectionsHeaderWrapper = (props) => {
  let newProps = Object.assign({}, defaultProps, props)
  const ColorCollectionsWrapped = CollectionsHeaderWrapper(ColorCollections)
  return shallow(<ColorCollectionsWrapped {...newProps} />)
}

describe('CollectionsHeaderWrapper', () => {
  let collectionsHeaderWrapper
  beforeAll(() => {
    if (!collectionsHeaderWrapper) {
      collectionsHeaderWrapper = getCollectionsHeaderWrapper()
    }
  })

  it('should match snapshot', () => {
    expect(collectionsHeaderWrapper).toMatchSnapshot()
  })

  it('should render Link and with to prop as /active', () => {
    expect(collectionsHeaderWrapper.find(Link).exists()).toBe(true)
    expect(collectionsHeaderWrapper.find(Link).prop('to')).toEqual('/active')
  })

  it('should render CLOSE button', () => {
    expect(collectionsHeaderWrapper.find(`button.${buttonRight}`).exists()).toBe(true)
  })

  it('should not render BACK button', () => {
    expect(collectionsHeaderWrapper.find(`button.${buttonLeft}`).exists()).toBe(false)
  })

  it('should render wrapped component', () => {
    expect(collectionsHeaderWrapper.find(ColorCollections).exists()).toBe(true)
  })
})

describe('CollectionsHeaderWrapper states & events', () => {
  let collectionsHeaderWrapper
  beforeAll(() => {
    if (!collectionsHeaderWrapper) {
      collectionsHeaderWrapper = getCollectionsHeaderWrapper()
    }
  })

  it('should update heading when setHeader is called on wrapped component', () => {
    const headerContent = 'Color Collections'
    collectionsHeaderWrapper.find(ColorCollections).props().setHeader(headerContent)
    expect(collectionsHeaderWrapper.find(`div.${heading}`).contains(headerContent)).toBe(true)
  })

  it('should render BACK button when isShowBack state is true', () => {
    collectionsHeaderWrapper.setState({ isShowBack: true })
    expect(collectionsHeaderWrapper.find(`button.${buttonLeft}`).exists()).toBe(true)
  })

  it('should update isShowBack state when showBack is called by wrapped component', () => {
    collectionsHeaderWrapper.find(ColorCollections).props().showBack()
    expect(collectionsHeaderWrapper.state('isShowBack')).toEqual(true)
  })

  it('should remove BACK button when it is clicked', () => {
    collectionsHeaderWrapper.setState({ isShowBack: true })
    collectionsHeaderWrapper.find(`button.${buttonLeft}`).simulate('click')
    expect(collectionsHeaderWrapper.find(`button.${buttonLeft}`).exists()).toBe(false)
  })
})

it('renders back button and header when initialized with collectionSummary as a property', () => {
    const collectionsHeaderWrapper = getCollectionsHeaderWrapper({ collectionSummary: { name: '2016 Pura Vida' } })
    expect(collectionsHeaderWrapper.find(`button.${buttonLeft}`).exists()).toBe(true)
    expect(collectionsHeaderWrapper.find(`div.${heading}`).contains('2016 Pura Vida')).toBe(true)
})
