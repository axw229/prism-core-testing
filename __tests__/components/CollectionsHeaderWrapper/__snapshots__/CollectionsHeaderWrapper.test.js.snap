// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionsHeaderWrapper should match snapshot 1`] = `
<div
  className="collections-header__wrapper"
>
  <div
    className="collections-header__header"
  >
    <div
      className="collections-header__heading"
    />
    <Link
      to="/active"
    >
      <button
        className="collections-header__button collections-header__button--right"
      >
        <div
          className="collections-header__close"
        >
          <span>
            CLOSE
          </span>
          Â 
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Array [
                "fa",
                "chevron-up",
              ]
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </div>
        <div
          className="collections-header__cancel"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Array [
                "fa",
                "times",
              ]
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </div>
      </button>
    </Link>
  </div>
  <div
    className="collections-header__content"
  >
    <ColorCollections
      isShowBack={false}
      setHeader={[Function]}
      showBack={[Function]}
    />
  </div>
</div>
`;
